articles/115_topic_and_service_name_mapping.md

```diff
@@ -0,0 +1,317 @@
+---
+layout: default
+title: Topic and Service name mapping to DDS
+permalink: articles/topic_and_service_names.html
+abstract:
+  This article describes the proposed mapping between ROS topic and service names to DDS topic and service names.
+author: '[William Woodall](https://github.com/wjwwood)'
+published: true
+---
+
+- This will become a table of contents (this text will be scraped).
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+## Context
+
+Before proposing constraints for ROS 2 topic and service names and a mapping to underlying DDS topics, this article will first summarize the existing guidelines and restrictions for ROS 1 and DDS.
+
+### ROS 1 Names
+
+In ROS 1, topic and service name guidelines are covered all under the umbrella of "ROS Names" and have these restrictions:
+
+*From [http://wiki.ros.org/Names](http://wiki.ros.org/Names):*
+
+```
+A valid name has the following characteristics:
+
+* First character is an alpha character ([a-z|A-Z]), tilde (~) or forward slash (/)
+* Subsequent characters can be alphanumeric ([0-9|a-z|A-Z]), underscores (_), or forward slashes (/)
+
+Exception: base names (described below) cannot have forward slashes (/) or tildes (~) in them.
+```
+
+### DDS Topic Names
+
+In DDS, topic names are restricted by these restrictions:
+
+*From DDS 1.4 specification, or the [RTI documentation](http://community.rti.com/rti-doc/510/ndds/doc/html/api_cpp/group__DDSQueryAndFilterSyntaxModule.html):*
+
+```
+TOPICNAME - A topic name is an identifier for a topic, and is defined as any series of characters 'a', ..., 'z', 'A', ..., 'Z', '0', ..., '9', '_' but may not start with a digit.
+```
+
+*Note:* that the DDS specification has a known typo, where it says `-` are allowed, but the RTI documentation correctly lists `_` as allowed.
+
+Additionally, DDS has a hard limit on topic names of 256 characters, so an additional goal is to minimize the number of extra characters used when mapping from ROS to DDS names.
+
+## ROS 2 Topic and Service Name Constraints
+
+In this section an outline of the proposed constrains for ROS 2 topic and service names will be enumerated along with rationales where appropriate.
+
+For convenience here is a summary of all rules for topic and service names in ROS 2:
+
+- may contain alphanumeric characters (`[0-9|a-z|A-Z]`), underscores (`_`), tildes (`~`), or forward slashes (`/`)
+- must not start with a numeric character (`[0-9]`)
+- must not contain any number of repeated underscores (`_`)
+- must not end with an underscore (`_`)
+- must not have an underscore (`_`) proceeded by a forward slash (`/`), i.e. `_/`
+- must not end with a forward slash (`/`)
+- must not contain any number of repeated forward slashes (`/`)
+- must not have tilde (`~`) adjacent to anything other than forward slashes (`/`)
```

Comment:
I had that before, but my goal was to have each rule be correct without other rules (otherwise the rules are somewhat contradicting each other). For example, the valid name `~/foo` when applied to "May contain tilde (`~`) characters only if each individual tilde is surrounded by forward slashes" is invalid because the `~` is not surrounded by `/`.

Either way, it sounds like we'll be changing this so I'll wait till that discussion is done.